Integrate jobserver support to parallel codegen

alexcrichton opened the pull request:
This commit integrates the `jobserver` crate into the compiler. The crate was
previously integrated in to Cargo as part of rust-lang/cargo#4110. The purpose
here is to two-fold:

* Primarily the compiler can cooperate with Cargo on parallelism. When you run
  `cargo build -j4` then this'll make sure that the entire build process between
  Cargo/rustc won't use more than 4 cores, whereas today you'd get 4 rustc
  instances which may all try to spawn lots of threads.

* Secondarily rustc/Cargo can now integrate with a foreign GNU `make` jobserver.
  This means that if you call cargo/rustc from `make` or another
  jobserver-compatible implementation it'll use foreign parallelism settings
  instead of creating new ones locally.

As the number of parallel codegen instances in the compiler continues to grow
over time with the advent of incremental compilation it's expected that this'll
become more of a problem, so this is intended to nip concurrent concerns in the
bud by having all the tools to cooperate!

Note that while rustc has support for itself creating a jobserver it's far more
likely that rustc will always use the jobserver configured by Cargo. Cargo today
will now set a jobserver unconditionally for rustc to use.

rust-highfive commented:
r? @arielb1

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
r? @michaelwoerister

michaelwoerister commented on src/librustc_trans/back/write.rs:
```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+                       scope,
+                       tx.clone(),
+                       work_items.pop().unwrap(),
+                       work_items.len());
```
I'm not very fond this: mutating `work_items` via `pop` and then taking its `len`. I assume that we have a defined evaluation order of function arguments, but I don't like relying on it.

michaelwoerister commented on src/librustc_trans/back/write.rs:
```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+        // possible. Remember that we have an ambient token available to us
+        // hence the `+1` here.
+        //
+        // Also note that we may actually acquire more tokens than we need, so
```
When does that happen? If we abort early because of an error?

michaelwoerister commented on src/librustc_trans/back/write.rs:
```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+        //
+        // Also note that we may actually acquire more tokens than we need, so
+        // in that case just truncate the `tokens` list every time we pass
+        // through here.
```
Could you add that truncating implies dropping and thus releasing tokens?

michaelwoerister commented on src/librustc_trans/back/write.rs:
```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+                       work_items.len());
+            running += 1;
+        }
+        tokens.truncate(running.saturating_sub(1));
```
I'm not quite sure how this works. Can't this cause tokens to be lost without a `spawn_work` having been called for them?

michaelwoerister commented on src/librustc_trans/back/write.rs:
```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+
+        // Set up a destructor which will fire off a message that we're done as
+        // we exit.
+        struct Bomb {
```
We should have something like this in libstd.

michaelwoerister commented on src/librustc_trans/back/lto.rs:
```diff
@@ -82,16 +84,11 @@ pub fn run(sess: &session::Session,
     // module that we've got.
-    link::each_linked_rlib(sess, &mut |cnum, path| {
-        // `#![no_builtins]` crates don't participate in LTO.
-        if sess.cstore.is_no_builtins(cnum) {
```
I see, it's added in later again.

alexcrichton replied:
Yeah this query just ended up having a lot of dependencies on `sess` so I figured it'd be best to move it way up to the beginning instead of only running it back here.

michaelwoerister commented on src/librustc_trans/back/write.rs:
```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+        if sess.cstore.is_no_builtins(cnum) {
+            return
+        }
+        each_linked_rlib.push((cnum, path.to_path_buf()));
```
If the `each_linked_rlib` field is LTO-specific, we should probably change to the name to reflect this.

michaelwoerister commented on src/librustc_trans/back/write.rs:
```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+        // Execute the work itself, and if it finishes successfully then flag
+        // ourselves as a success as well.
+        if execute_work_item(&cgcx, work).is_err() {
+            drop(cgcx.tx.send(Message::AbortIfErrors));
```
One could argue that it would be cleaner to also `mem::forget` the `bomb` in this case.

alexcrichton replied:
Yeah I wasn't quite sure how this should be handled, I think that if you see a `FatalError` then a diagnostic has already been sent off, which in turn already sent `AbortIfErrors`. In that sense it may be fruitless to send another message here, so I'll just ignore the result.

michaelwoerister commented on src/librustc_trans/scope.rs:
```diff
@@ -0,0 +1,114 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Scoped threads, copied from `crossbeam`
```
Could we also use `crossbeam` directly?

alexcrichton replied:
Hm upon further inspection, I don't see why not!

alexcrichton commented:
Ok, updated! @michaelwoerister I added a large comment above the "main loop" which I believe should answer your questions about the token management, but if you'd like me to clarify anything please just let me know!

bors commented:
:umbrella: The latest upstream changes (presumably #42676) made this pull request unmergeable. Please resolve the merge conflicts.

michaelwoerister commented on src/librustc_trans/back/write.rs:
```diff
@@ -1033,94 +956,283 @@ fn execute_work_item(cgcx: &CodegenContext,
+    // manner we can ensure that the maximum number of parallel workers is
+    // capped at any one point in time.
+    //
+    // The jobserver protocol is a little unique, however. We, as a running
```
Because concurrent programming isn't complicated enough by itself already :stuck_out_tongue:

michaelwoerister commented:
Thanks for the clarifying comment about the jobserver protocol!

r=me once the merge conflict is fixed.

alexcrichton commented:
@bors: r=michaelwoerister

bors commented:
:pushpin: Commit 5d00e5e has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 5d00e5ea2892d376aa18f8db2e4db435e097a81a -->

bors commented:
:hourglass: Testing commit 5d00e5ea2892d376aa18f8db2e4db435e097a81a with merge 6cb3b992db1aa5c9952c187dca9b30ec0c3d98d4...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3644)

alexcrichton commented:
@bors: r=michaelwoerister

bors commented:
:pushpin: Commit a014634 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister a014634c1a0ee939207dcd1e8d64dfcd0ebec586 -->

bors commented:
:hourglass: Testing commit a014634c1a0ee939207dcd1e8d64dfcd0ebec586 with merge 016496955016d5d75d27180c10a158aad0083c8d...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3647)

Mark-Simulacrum commented:
Linking failure?

```
[01:33:02] error: linking with `link.exe` failed: exit code: 1120
[01:33:02]   |
[01:33:02]   = note: "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\bin\\amd64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LIBPATH:C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "
C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-pass\\smallest-hello-world.0.o" "/OUT:C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-pass\\smallest-hello-world.stage2-x86_64-pc-windows-msvc.exe" "/OPT:REF,ICF" "
/DEBUG" "/LIBPATH:C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-pass" "/LIBPATH:C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-pass\\smallest-hello-world.stage2-x86_64-pc-windows-msvc.run-pass.libaux" "/LIBPAT
H:C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\native\\rust-test-helpers" "/LIBPATH:C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\projects\\rust\\build\\x86_64-pc-windows-
msvc\\stage2\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc_system-ace597b8fd7407ce.rlib" "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-216e0f775a886879.rlib"
[01:33:02]   = note: smallest-hello-world.0.o : error LNK2019: unresolved external symbol puts referenced in function main
[01:33:02]           LINK : error LNK2001: unresolved external symbol mainCRTStartup
[01:33:02]           C:\projects\rust\build\x86_64-pc-windows-msvc\test\run-pass\smallest-hello-world.stage2-x86_64-pc-windows-msvc.exe : fatal error LNK1120: 2 unresolved externals
[01:33:02]
[01:33:02]
[01:33:02] error: aborting due to previous error(s)
[01:33:02]
```

alexcrichton commented:
@bors: r=michaelwoerister

bors commented:
:pushpin: Commit 2376e0d has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 2376e0d5b3671988b225f940149945d64978b5f9 -->

bors commented:
:lock: Merge conflict

bors commented:
:umbrella: The latest upstream changes (presumably #42664) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
@bors: r=michaelwoerister

bors commented:
:pushpin: Commit 451d392 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 451d39275cd95096548143151041df198ef3c9bb -->

alexcrichton commented:
@bors: r=michaelwoerister

bors commented:
:pushpin: Commit 201f069 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 201f06988fb77c9865b0ddc60a8d7b5701c6dbbe -->

bors commented:
:hourglass: Testing commit 201f06988fb77c9865b0ddc60a8d7b5701c6dbbe with merge 694adee9e284600f4aaafaf03d442ec1f57f46fd...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/245482773?utm_source=github_status&utm_medium=notification)

alexcrichton commented:
@bors: retry

* osx timed out

On Wed, Jun 21, 2017 at 4:24 PM, bors <notifications@github.com> wrote:

> 💔 Test failed - status-travis
> <https://travis-ci.org/rust-lang/rust/builds/245482773?utm_source=github_status&utm_medium=notification>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/42682#issuecomment-310209507>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95FEbgYHMXulmMHTszBVJR6hpkE9Lks5sGYoOgaJpZM4N7l7L>
> .
>

bors commented:
:hourglass: Testing commit 201f06988fb77c9865b0ddc60a8d7b5701c6dbbe with merge 80271e8edfb188cfb9e2f02ac26136b4f58fbef0...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3659), [status-travis](https://travis-ci.org/rust-lang/rust/builds/245595870?utm_source=github_status&utm_medium=notification)
Approved by: michaelwoerister
Pushing 80271e8edfb188cfb9e2f02ac26136b4f58fbef0 to master...

jdm commented:
So does this only support invoking `cargo`/`rustc` from `make`, but the behaviour of invoking `make` from a Cargo build script is unchanged?

alexcrichton commented:
@jdm it's a little more nuanced than that. Cargo also *creates* a jobserver in addition to consuming one, meaning that rustc will basically always use that jobserver now. If Cargo inherits a jobserver though then rustc likely will too.

You need to tweak makefiles calling rustc/cargo though to actually let them inherit the jobserver, notably adding a `+` to the beginning of the rule definition.

For build scripts invoking make the `make` subprocess will inherit Cargo's jobserver if no `-j` argument is passed, but if `-jN` is passed then that'll override the inherited jobserver.
